File: NeetCode/LinkedList/Medium/61.RotateList.py
# Given the head of a linked list, rotate the list to the right by k places.

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def rotateRight(head, k):
    if not head:
        return head

    length, end_node = 1, head
    while end_node.next:
        length += 1
        end_node = end_node.next

    if k % length == 0:
        return head

    current = head
    for i in range(length - k % length - 1):
        current = current.next

    new_node = current.next
    end_node.next = head
    current.next = None

    return new_node

File: swift/Neetcode/LinkedList/Medium/61.RotateList.swift
// Given the head of a linked list, rotate the list to the right by k places.

func rotateRight(_ head: ListNode?, _ k: Int) -> ListNode? {
    if head == nil { return nil }

    var length = 1, end = head

    while end?.next != nil {
        length += 1
        end = end?.next
    }

    if k % length == 0 { return head }

    let n = length - k % length - 1     
    var current = head

    for _ in 0..<n {
        current = current?.next
    }

    var newHead = current?.next
    end?.next = head
    current?.next = nil

    return newHead
}